tests/test_parser.py

```diff
@@ -257,6 +259,98 @@ def test_convert_to_date_cast_result_to_date(self, mocked_locale,
                          expected_output)
 
 
+class Test_bs_safeget(unittest.TestCase):
+    def test_bs_safeget_return_expected_text(self):
+        mocked_soup = MagicMock()
+        selector = 'some_tag'
+        mocked_soup.select.return_value = ['some text']
+        self.assertEqual(bs_safeget(mocked_soup, selector), 'some text')
+
+    def test_bs_safeget_return_false_when_selector_is_missing(
+            self):
+        mocked_soup = Mock()
+        mocked_soup.select.return_value = ''
+        selector = 'h1'
+        self.assertFalse(bs_safeget(mocked_soup, selector))
+
+
+@patch('smarsy.parse.bs_safeget')
+class TestCreateParentsDict(unittest.TestCase):
+    def test_bs_safeget_called_with_expected_html_for_img_getting(
+            self, mocked_get):
+        tag1, tag2 = '[valign=top]', 'img[src]'
+        create_parents_dict('html')
+        mocked_get.assert_called_with('html', tag1, tag2)
+
+    def test_create_parents_dict_return_false_with_none_img(self, mocked_get):
+        mocked_get.return_value = None
+        self.assertFalse(create_parents_dict('html'))
+
+    @patch('builtins.hasattr')
+    def test_bs_safeget_return_expected_img(self, mock_hasattr, mocked_get):
+        mocked_get.return_value = 'some_img.jpg'
+        create_parents_dict('html')
+        self.assertTrue(mock_hasattr.call_count, 1)
+
+
+class TestParseParentPage(unittest.TestCase):
+    @patch('smarsy.parse.BeautifulSoup')
+    def test_parent_page_content_called_with_expected_html(self, mocked_soup):
+        html = '<tr></tr>'
+        parent_page_content_to_object(html)
+        mocked_soup.assert_called_with(html, 'html.parser')
+
+    @patch('smarsy.parse.BeautifulSoup', side_effect=TypeError)
+    def test_parent_page_content_raise_exeption_with_wrong_file_format(
+            self, mocked_soup):
+        html = 12344
+        with self.assertRaises(TypeError) as error:
+            parent_page_content_to_object(html)
+        self.assertEqual('Wrong file format', str(error.exception))
+
+    @patch('smarsy.parse.BeautifulSoup.select', return_value='')
+    def test_parent_page_content_return_false_with_wrong_parent_tab(
+            self, mocked_soup_select):
+        html = '<tr></tr>'
+        parent_page_content_to_object(html)
```

Comment:
why is it called twice?

Reply:
Not twice in first test mocked @patch('smarsy.parse.BeautifulSoup'), in second @patch('smarsy.parse.BeautifulSoup.select', return_value='')

Reply:
315 line - first time,  316 - 2nd

Reply:
Line 315 deleted